Lesson9

sovetnik opened the pull request.

sovetnik committed:
Votes for question

Vote model
Concern Votable
Question include votable

question voting work in browser

Question voting feature spec

split #vote action

Voting now has actions #good & #shit,
which directly calls try_increase & try_decrease on Vote

Answers voting

Vote model not care about user

disabled state for votable moved in helper
access to voting check in controller

Voting JS fix

full test cover for vote

set votable moved in module
shared_example spec for Vote model & controller
added more voting examples

palkan commented on app/controllers/concerns/voting.rb:
```diff
@@ -0,0 +1,30 @@
+  extend ActiveSupport::Concern
+
+  included do
+    before_action :set_votable, :load_vote, only: [:good, :shit]
```
А где же метод `set_votable`?  Он должен определяться в самом модуле, иначе смысл модуля теряется.

sovetnik replied:
Сделано.

palkan commented on app/models/concerns/votable.rb:
```diff
@@ -0,0 +1,19 @@
+    votes.find_or_initialize_by(voter: user)
+  end
+
+  def rating
```
Вообще, рейтинг - это сумма голосов по всем пользователям, а не голос конкретного пользователя.

palkan commented on app/models/vote.rb:
```diff
@@ -0,0 +1,51 @@
+  protected
+
+  def overall
+    Vote.find_or_initialize_by(votable: votable, voter: votable)
```
Во-первых, это чушь (`voter: votable`???).
Во-вторых, зачем это? Увеличиваем один голос, а затем увеличиваем другой (хотя вроде как тот же)??

sovetnik replied:
Это не тот же, Vote.find_or_initialize_by(votable: votable, voter: votable) - это общий рейтинг вопроса или ответа.

palkan replied:
Общий рейтинг вопроса/ответа хранится отдельной записью? Это неверно.
Рейтинг - это просто поле в модели вопроса/ответа.

sovetnik replied:
Если он будет отдельной записью, то ее нужно будет создавать в двух местах, и подключение Votable будет требовать миграций.

palkan replied:
Миграции - это не страшно. Лишние запросы для выгрузки/создания записи-голоса - вот это проблема.
Почему бы тогда все поля не хранить в отдельных таблицах? Такой метод часто применяют в CMS, чтобы _домохозяйкам_ не надо было работать с БД.
Если рейтинг приналежит объекту и только ему - значит он должен _лежать_ в базе данных вместо с самим объектом.

То, что ты пытаешься сделать - это пример _перенормализации_. В лучшем случае для выгрузки списка ответов с рейтингами тебе придется делать JOIN, в худшем - сделаешь N+1 запрос.

sovetnik replied:
Я выбирал из нескольких вариантов, и выбирал по нескольким критериям, вот вчера даже в чате с народом этот путь обсуждали.
1) никто напрямую не видит объект рейтинга и не может на него влиять
2) изменение происходит через объект голоса, принимающий значения от -1 до 1
3) при удалении вопроса или ответа удаляются все записи голосов и рейтинга связанные с ним через votable
4) при удалении юзера, удаляются его голоса, но это не изменяет рейтинг.
5) подключаться Votable должно одной строкой
6) у подключившего есть метод rating, возвращающий рейтинг (число)
7) сортировка ответов к вопросу по рейтингу возможна

Попробовал отсортировать, действительно выглядит длинно:
Answer.where(id: (Vote.where(voter: Question.last.answers).order(rate: :desc).pluck :votable_id))
  Question Load (0.5ms)  SELECT  "questions".\* FROM "questions"  ORDER BY "questions"."id" DESC LIMIT 1
   (0.7ms)  SELECT "votes"."votable_id" FROM "votes" WHERE "votes"."voter_type" = 'Answer' AND "votes"."voter_id" IN (SELECT "answers"."id" FROM "answers" WHERE "answers"."question_id" = $1)  ORDER BY "votes"."rate" DESC  [["question_id", 12]]
Answer Load (0.5ms)  SELECT "answers".\* FROM "answers" WHERE "answers"."id" IN (212, 214, 211)

Сделаю миграции. Получается, это нормальная практика - добавлять к моделям поля, которые нужны только для concern?

sovetnik replied:
> > Почему бы тогда все поля не хранить в отдельных таблицах? 
> > я как-то месяц провел изучая графовые базы, отдельно останавливался на neo4j с ее языком Cypher, но она тогда с рельсами не оч хорошо работала, gem orm был альфа версией. Мы их будем в курсе касаться?

palkan replied:
Графовых баз в курсе не будет, это тема совсем не для junior dev, и даже не для middle.

Добавлять поля к моделям при добавлении новых свойств - это нормальная практика.
Например, стандартный рельсовый counter_cache (по сути аналог рейтинга, только для количества, а не суммы) требует отдельной колонки.

Предлагаю рассмотреть довольно распространненую ситуацию.
Рассмотрим стандартую модель User(id, email).
Со временем мы захотим добавить больше информации о пользователе: фамилия и имя.
Эти свойства тесно связаны с пользователем и используются повсеместно в системе.
Добавляем для них новые поля в модель.
Затем нам понадобились другие данные о пользователе: телефон, адрес, индекс, кличка домашнего животного.
Добавлять новые поля или создавать новую модель?
Ответ на этот вопрос зависит от приложения.
Если все, что делает приложение - это выводит анкету пользователя со всеми этими данными, то логично добавлять новые поля и выгружать всю информацию сразу.
Если же анкетные данные используются только, например, в информации о профиле пользователя, а в большинстве случаев нам нужно только имя/фамилия (как в нашем проекте, чтобы подписывать вопросы/ответы), то нужно выносить анкетные данные в отдельную модель и делать связь `User has_one :profile`.

Есть еще альтернативный вариант: создать schemeless поле в модели User (json, hstore) и хранить подобные данные в нем.

sovetnik replied:
Про hstore в курсе, только что оно есть. почитаю подробнее обязательно.

Вопрос появился про increment! и валидации:
В модели Vote:

``` ruby
validates :rate, numericality: { only_integer: true,
                                   greater_than_or_equal_to: -1,
                                   less_than_or_equal_to: 1 }
  def increase
    transaction do
      self.increment! :rate
      votable.increment! :rating
    end
  end 
```

в доках http://apidock.com/rails/ActiveRecord/Persistence/increment%21
если делаю не bang, он обновляет только votable, если bang, то обоих и vote.valid? становится false, но в базу сохраняется. есть ли способ заставить increment учитывать валидацию или вернуть как было?

``` ruby
  def increase
    if rate < 1
      transaction do
        self.increment! :rate
       votable.increment! :rating
      end
    else
      false
    end
  end
```

palkan commented on app/models/vote.rb:
```diff
@@ -0,0 +1,51 @@
+  end
+
+  def increase
+    self.rate += 1
```
Для этого придумали функцию `ActiveRecord#increment!`.

sovetnik replied:
Заменил.

palkan commented:
В тестах контроллеров недостаточно только одной проверки возвращаемых статусов.
Нам важней проверять, что создаются (или не создаются) голоса, причем корректные.

oblaka commented:
Вов, а какими критериями нужно оценивать мою реализацию рейтинга, чтобы придти к выводу, что это неверно?

oblaka commented:
Вообще, я предполагал в будущем брать эти цифры из voter.power или типа того, когда появятся рейтинги.

sovetnik committed:
rating moved in votable models

Votes on callback & increment

cast and revoke votes by callbacks
model, controller & tests done
change polymorphic voter to user

Vote & Votable fix

Voting concern fix

votes fix

add validation & index in order to have uniq votes

Voting Feature

added link to revoke on answers & questions
added feature specs for answers & questions voting

palkan commented on app/models/vote.rb:
```diff
@@ -0,0 +1,22 @@
+  delegate :rating, to: :votable
+
+  def increase
+    self.rate += 1
```
Это надо делать через `increment!`, в противном случае при одновременном обновлении из двух потоков результат может быть неверным.

sovetnik replied:
я про increment! спрашивал, как его с валидацией подружить, может, есть способ?
в нынешнем виде могу завернуть в with_lock, или это тоже неверно?

palkan commented on app/models/vote.rb:
```diff
@@ -0,0 +1,22 @@
+
+  def increase
+    self.rate += 1
+    save ? votable.increment!(:rating) : false
```
А для этого существуют коллбеки (например, `after_save`).

palkan commented on app/models/vote.rb:
```diff
@@ -0,0 +1,22 @@
+  end
+
+  def decrease
+    self.rate -= 1
```
Получается, что если я сначала проголосовал "за", а потом "против", то в базе останется голос с нулем?
Это плохо, мы зачем-то храним _бессмысленный_ объект.

Более правильная схема такая:
- пользователь голосует "за" - создается голос со значением 1;
- пользователь не может голосовать, пока не отменит голос;
- пользователь отменяет голос - голос удаляется из базы;
- теперь пользователь может проголосовать "против" - создается голос со значением -1.

Кроме того, это упростит изменение рейтинга голосуемого объекта, можно будет обойтись без лишних методов и использовать коллбеки на создание и удаление.

sovetnik replied:
я планировал наделить этот объект смыслом, вроде: переголосовать можно только один раз в день, например. или только пользователь с некими правами может переголосовать.

palkan commented on spec/features/questions/vote_spec.rb:
```diff
@@ -0,0 +1,80 @@
+      expect(page).to have_css("#shit_#{question.class.name.downcase}_#{question.id}.disabled")
+      click_on('good!')
+      within "#rating_#{question.class.name.downcase}_#{question.id}" do
+        expect(page).to have_content('0')
```
Вот такой ситуации быть не должно: если я проголосовал "против", а потом передумал и проголосовал "за", рейтинг должен оказаться равным 1, а не нулю. По сути, я _переголосовал_ - отменил свой первый голос и создал новый.

palkan commented on app/models/vote.rb:
```diff
@@ -0,0 +1,33 @@
+  before_destroy :revoke
+
+  def increase
+    self.increment! :rate unless persisted?
```
Очень странное и неочевидное условие.
Таким образом ты пытаешься запретить изменять голос после создания?

sovetnik replied:
да, именно. уже придумал, как его избежать иным способом.

palkan commented on app/models/vote.rb:
```diff
@@ -0,0 +1,33 @@
+  end
+
+  def revoke
+    votable.decrement! :rating, rate unless new_record?
```
`after_destroy` не будет вызываться для `new_record?`

sovetnik replied:
да, в консоли пробовал уничтожить новый объект, и метод вызывался.

palkan commented on app/models/concerns/votable.rb:
```diff
@@ -0,0 +1,15 @@
+  extend ActiveSupport::Concern
+
+  included do
+    has_many :votes, as: :votable, dependent: :destroy
```
Здесь лучше использовать `dependent: :delete`, нам не нужно вызывать коллбеки на удаленных голосах.

palkan commented on spec/controllers/shared/voting_example.rb:
```diff
@@ -0,0 +1,242 @@
+        end
+      end
+
+      context 'vote already exists and positive' do
```
Вот эти два случая - "exists and positive" и "exists and negative"  - суть одно и то же: если голос уже есть, то ничего не меняется.

Это сразу наводит на мысль, что именно такая проверка и должна быть при голосовании, а не  `unless persisted?`.

palkan commented:
Еще мне не нравится то, что в консерне голосования основной действующий объект - Голос, а не  Голосуемое.
Во-первых, это нарушает соглашение о том, что контроллер _работает_ с одним ресурсом.
Во-вторых, это усиливает связность нашего кода: мы привязываемся в конкретной реализации голосования.

Вот, как мне кажется, правильный вариант:

```ruby
module Voting
  def good
     @votable.vote_up(current_user)
  end

  def shit
     @votable.vote_down(current_user)
  end

  def cancel_vote
     @votable.cancel_vote(current_user)
  end
end
```

При таком подходе нам совершенно не важно, что происходит внутри модели, как именно реализовано голосования. Это нам дает:
- более простое тестирование (весь функционал можно проверить на уровне модели);
- более простой рефакторинг в будущем;
- весь _рабочий_ код голосования будет вынесен в модуль Votable.
